hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/LatencyTracker.java

```diff
@@ -0,0 +1,215 @@
+  private final ConcurrentLinkedQueue<String> queue = new ConcurrentLinkedQueue<String>();
+
+  // whether the latency tracker has been enabled
+  private boolean enabled = false;
+
+  // the host name
+  private String hostName;
+
+  // the file system name
+  private String filesystemName;
+
+  // the account name
+  private String accountName;
+
+  // singleton latency reporting format
+  private String singletonLatencyReportingFormat;
+
+  // aggregate latency reporting format
+  private String aggregateLatencyReportingFormat;
+
+  public LatencyTracker(String filesystemName, String accountName, AbfsConfiguration configuration) {
+    this(filesystemName, accountName, configuration.shouldTrackLatency());
+  }
+
+  protected LatencyTracker(String filesystemName, String accountName, boolean enabled) {
+    this.enabled = enabled;
+    this.filesystemName = filesystemName;
+    this.accountName = accountName;
+
+    LOG.debug("LatencyTracker configuration: {}", enabled);
+
+    if (enabled) {
+      try {
+        hostName = InetAddress.getLocalHost().getHostName();
+      } catch (UnknownHostException e) {
+        hostName = "UnknownHost";
+      }
+
+      singletonLatencyReportingFormat = "h=" + hostName + " t=%s a=" + accountName + " c=" + filesystemName + " cr=%s ce=%s r=%s l=%s%s";
+      aggregateLatencyReportingFormat = "h=" + hostName + " t=%s a=" + accountName + " c=" + filesystemName + " cr=%s ce=%s r=%s l=%s ls=%s lc=%s%s";
+    }
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          String callerName,
+          String calleeName,
+          boolean success,
+          AbfsHttpOperation res) {
+    if (!enabled) {
+      return;
+    }
+
+    Instant operationStop = getLatencyInstant();
+
+    recordClientLatency(operationStart, operationStop, callerName, calleeName, success, res);
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          Instant operationStop,
+          String callerName,
+          String calleeName,
+          boolean success,
+          AbfsHttpOperation res) {
+    if (!enabled) {
+      return;
+    }
+
+    Instant trackerStart = Instant.now();
+    long latency = isValidInstant(operationStart) && isValidInstant(operationStop)
+            ? Duration.between(operationStart, operationStop).toMillis() : -1;
+
+    String latencyDetails = String.format(singletonLatencyReportingFormat,
+            Instant.now(),
+            callerName,
+            calleeName,
+            success ? "Succeeded" : "Failed",
+            latency,
+            res == null ? "" : (" " + res.toKvpString()));
+
+    this.offerToQueue(trackerStart, latencyDetails);
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          String callerName,
+          String calleeName,
+          boolean success,
+          Instant aggregateStart,
+          long aggregateCount,
+          AbfsHttpOperation res) {
```

Comment:
I think this is only ABFS specific field. Can we make this generic and be able to use it by others?

Reply:
Abstracted this out into an interface.